bip-signmessage-redone.mediawiki

```diff
@@ -0,0 +1,156 @@
+<pre>
+  BIP: ????
+  Layer: Applications
+  Title: Generic Signed Message Format
+  Author: Karl-Johan Alm <karljohan-alm@garage.co.jp>
+  Comments-Summary: No comments yet.
+  Comments-URI: https://github.com/bitcoin/bips/wiki/Comments:BIP-????
+  Status: Draft
+  Type: Standards Track
+  Created: 2020-10-01
+  License: CC0-1.0
+  Replaces: 322
+</pre>
+
+== Abstract ==
+
+A standard for interoperable generic signed messages based on the Bitcoin Script format.
+
+== Background ==
+
+* Assume two actors, a prover <code>P</code> and a verifier <code>V</code>.
+* <code>P</code> wants to prove that they own the private key <code>k</code> associated with a given address <code>A</code> (which in turn is derived from the pubkey <code>kG</code>).
+* Let <code>V</code> generate a message <code>M</code> and hand this to <code>P</code>.
+* <code>P</code> generates a signature <code>S</code> by signing the message <code>M</code> using <code>k</code>. Given <code>S</code>, <code>V</code> can check that <code>P</code> has the private key associated with <code>A</code>.
+
+The astute reader will notice that the above is missing a critical part, namely the pubkey <code>kG</code>, without which the verifier cannot actually verify the message. The current message signing standard solves this via a cryptographic trick, wherein the signature <code>S</code> above is a special "recoverable signature" type. Given the message <code>M</code> and the signature <code>S</code>, it is then possible to recover the pubkey <code>kG</code>. The system thus derives the address for the pubkey <code>kG</code>, and if it does not match <code>A</code>, the proof is deemed invalid.
+
+While this is a neat trick, it unnecessarily restricts and complicates the message signing mechanism; for instance, it is currently not possible to sign a message for a P2SH address, because there is no pubkey to recover from the resulting signature.
+
+== Motivation ==
+
+The current message signing standard only works for P2PKH (1...) addresses. By extending it to use a Bitcoin Script based approach, it could be made more generic without causing a too big burden on implementers, who most likely have access to Bitcoin Script interpreters already.
+
+== Specification ==
+
+This BIP follows the specification of BIP-325 challenges and solutions.
+
+Let there be two virtual transactions to_spend and to_sign.
+
+The "to_spend" transaction is:
+
+    nVersion = 0
+    nLockTime = 0
+    vin[0].prevout.hash = 0000...000
+    vin[0].prevout.n = 0xFFFFFFFF
+    vin[0].nSequence = 0
+    vin[0].scriptSig = OP_0 PUSH32[ message_hash ]
```

Comment:
The `SignatureHash` function in Core does not take anything from scriptSig. It takes the `scriptCode` as the supplied argument, and later uses it in the hashing.

AFAIU this `scriptSig = OP_0 PUSH32[ message_hash ]` is intended to convey that the message hash will be committed to as part of the script supplied to `SignatureHash`. But as `SignatureHash` does not use the contents of `scriptSig` in the input at all, and only uses the script that is supplied as argument, I believe that it would be more clear to state this explicitly. This is what I meant in an earlier message. The document should just specify all the relevant arguments to `SignatureHash`, roughly: `SignatureHash(CScript([OP_0, message_hash]), tx_to_spend, 0 /* nIn */, SIGHASH_ALL, 0 /* amount */)`

Reply:
(edited the above message, added sigversion as the parameter to SignatureHash)

The way to calculate the hash that will be signed need to be unambiguous. If we choose to use `SigVersion::WITNESS_V0` for sigversion, then `amount` should be fixed, because it will affect the hash.

Reply:
We do not actually sign this transaction, we sign the transaction spending it.